### ISSUE TITLE
Pyright tripping on tuple unpacking in async gen

### ISSUE THREAD

**Issue by betafcc** [microsoft/pyright#4028]
**Describe the bug**
Admittedly corner use case with easy workaround, but may be symptomatic:

**To Reproduce**

The following is annotated with pyright 1.1.274 output, note the unpacked values from `self.enumerate()` in `wrong` method:

```python
from __future__ import annotations

from typing import AsyncIterable, AsyncIterator, Generic, TypeVar

A = TypeVar("A")


class Aiter(Generic[A]):
    def __init__(self, fa: AsyncIterable[A]) -> None:
        self.fa = fa

    def __aiter__(self) -> AsyncIterator[A]:
        return self.fa.__aiter__()

    def enumerate(self) -> Aiter[tuple[int, A]]:
        async def gen():
            i = -1
            async for a in self:
                i += 1
                yield i, a

        return Aiter(gen())

    async def wrong(self) -> AsyncIterable[A]:
        async for i, value in self.enumerate():
            reveal_type(i)  # Type of "i" is "int"
            reveal_type(value)  # Type of "value" is "tuple[int, A@Aiter]"
            yield value  # Expression of type "tuple[int, A@Aiter]" cannot be assigned to yield type "A@Aiter"
```

Pyright does recognize the `i` as `int` but thinks `value` still is `tuple[int, A]`

Pyright 1.1.274 cli giving the unexpected output:

```sh
❱ npx -y pyright --pythonversion 3.10 dirty_bits/helpers/aitertools.py
No configuration file found.
pyproject.toml file found at /Users/betani/moteefe/dirty-bits.
Loading pyproject.toml file at /Users/betani/moteefe/dirty-bits/pyproject.toml
Pyproject file "/Users/betani/moteefe/dirty-bits/pyproject.toml" is missing "[tool.pyright]" section.
stubPath /Users/betani/moteefe/dirty-bits/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
pyright 1.1.274
/Users/betani/moteefe/dirty-bits/dirty_bits/helpers/aitertools.py
  /Users/betani/moteefe/dirty-bits/dirty_bits/helpers/aitertools.py:26:25 - information: Type of "i" is "int"
  /Users/betani/moteefe/dirty-bits/dirty_bits/helpers/aitertools.py:27:25 - information: Type of "value" is "tuple[int, A@Aiter]"
  /Users/betani/moteefe/dirty-bits/dirty_bits/helpers/aitertools.py:28:19 - error: Expression of type "tuple[int, A@Aiter]" cannot be assigned to yield type "A@Aiter"
    Type "tuple[int, A@Aiter]" cannot be assigned to type "A@Aiter" (reportGeneralTypeIssues)
  /Users/betani/moteefe/dirty-bits/dirty_bits/helpers/aitertools.py:33:21 - information: Type of "i" is "int"
  /Users/betani/moteefe/dirty-bits/dirty_bits/helpers/aitertools.py:34:21 - information: Type of "value" is "A@right"
1 error, 0 warnings, 4 informations 
Completed in 0.492sec
```

Mypy 0.982 giving the expected output:

```sh
❱ mypy dirty_bits/helpers/aitertools.py
dirty_bits/helpers/aitertools.py:26: note: Revealed type is "builtins.int"
dirty_bits/helpers/aitertools.py:27: note: Revealed type is "A`1"
dirty_bits/helpers/aitertools.py:33: note: Revealed type is "builtins.int"
dirty_bits/helpers/aitertools.py:34: note: Revealed type is "A`-1"
Success: no issues found in 1 source file
```

Besided the obvious `cast(A, value)` workaround, what I found weird is that pyright gets it right when defining an outer function:

```python
async def right(fa: Aiter[A]) -> AsyncIterable[A]:
    async for i, value in fa.enumerate():
        reveal_type(i)  # Type of "i" is "int"
        reveal_type(value)  # Type of "value" is "A@right"
        yield value # ok, type A as annotated
```

Therefore seems to be worth the ticket, looks like something specific with using from `self`

**Comment by betafcc** [comment]
Just noticed the same occurs in non-async Iterable:

```python
from __future__ import annotations

from typing import Iterable, Iterator, Generic, TypeVar

A = TypeVar("A")


class Iter(Generic[A]):
    def __init__(self, fa: Iterable[A]) -> None:
        self.fa = fa

    def __iter__(self) -> Iterator[A]:
        return self.fa.__iter__()

    def enumerate(self) -> Iter[tuple[int, A]]:
        def gen():
            i = -1
            for a in self:
                i += 1
                yield i, a

        return Iter(gen())

    def wrong(self) -> Iterable[A]:
        for i, value in self.enumerate():
            reveal_type(i)  # Type of "i" is "int"
            reveal_type(value)  # Type of "value" is "tuple[int, A@Iter]"
            yield value  # Expression of type "tuple[int, A@Iter]" cannot be assigned to yield type "A@Iter"


def right(fa: Iter[A]) -> Iterable[A]:
    for i, value in fa.enumerate():
        reveal_type(i)  # Type of "i" is "int"
        reveal_type(value)  # Type of "value" is "A@right"
        yield value  # ok, type A as annotated
```

**Comment by erictraut** [comment]
Here's a simplified repro that doesn't involve a generator.

```python
from typing import Iterator, Generic, TypeVar

A = TypeVar("A")

class Iter(Generic[A]):
    def __iter__(self) -> Iterator[A]:
        ...

    def enumerate(self) -> "Iter[tuple[int, A]]":
        ...

    def method1(self):
        for x in self.enumerate():
            reveal_type(x, expected_text="tuple[int, A@Iter")
```

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.279, which I just published. It will also be included in a future version of pylance.
